### DISCUSSION TITLE
How can I use WatchFiles to get a callback for my specific file while running nicegui UI?

### DISCUSSION THREAD

**Discussion by smojef** [zauberzeug/nicegui#417] [Q&A] [answered]
I am not sure how to deal with the ui.run() part.  I want my callback to be thread safe for nicegui calls.

To be more specific, I want to create a nicegui app for editing a file.  I will have a Save button to update the file. But I want to be notified about external changes to the file to prompt the user for reload.  I don't want to be notified about saves from my nicegui app.  

I have seen that nicegui already uses the WatchFiles module.

Thanks,

**Comment by falkoschindler** [comment, marked as answer]
With the `ui.run` parameters `uvicorn_reload_dirs`, `uvicorn_reload_includes`, and `uvicorn_reload_excludes` you should be able to limit NiceGUI's file watcher so that a second watcher does not interfere:

```py
import watchfiles
from nicegui import app, events, ui

async def watch_my_file():
    async for change in watchfiles.awatch('my_file.txt'):
        print(change, flush=True)
        ui.notify('File changed!')

ui.timer(0, watch_my_file, once=True)

ui.run(uvicorn_reload_excludes='my_file.txt')
```

The `ui.timer` is a little trick to run the async `watch_my_file()` in a UI context so that `ui.notify` knows the corresponding client.

**Reply by smojef** [reply]
Great! Thanks!

So if my_file.txt is not in the same folder as the script, I believe I don't need to exclude anything.

**Reply by smojef** [reply]
@falkoschindler, about the `ui.timer` trick, in the beginning, I was always using a `ui.timer(once=True)` to render the main components of my UI. This allowed me to use things such as `ui.notify` (requiring the UI to be running).  However, on one occasion, it did not load and I was left with an blank page.  So in the fear of losing such events in the future, I stop using this trick.  

Was that a bug?  Is there a mechanism in place to make sure that single timer events are actually delivered and processed (never missed)?

With your comment above I was thinking maybe I could use a repeating timer instead and deactivate it only once `watch_my_file` was actually called (like a watchdog).  Is this precaution necessary?

**Reply by falkoschindler** [reply]
There used to be an issue with `ui.timer` callbacks executing too early before the client is connected. But in version 1.0.9 we changed `ui.timer` to wait for a client connection before executing the callback for the first time. (#206) So every timer event should be delivered to the client.
